### ISSUE TITLE
Enable vi-insert-mode automatically on input fields

### ISSUE THREAD

**Issue by starryveshch** [atlas-engineer/nyxt#1462]
I find myself typing in an input field and wondering why my keys don't work all of a sudden then I find out I forgot to switch to vi-insert-mode. So I think it would be easier if this happens automatically,
* When opening prompt buffer.
* After selecting an input area or a text area through element-hint-mode.
#### What I tried
I tried enabling vi-insert-mode automatically in the prompt buffer by hooking it to prompt-buffer-mode-after-hook, but it only turns on vi-insert-mode globally with the prompt still on vi-normal-mode.
#### Related issues
* This issue #1461 talks about how the "g" key is not usable. This is because the user is on vi-normal-mode.
* This issue #739 talks about not forwarding unbound keys to avoid people typing in vi-normal-mode.

**Comment by edgar-vincent** [comment]
> I tried enabling vi-insert-mode automatically in the prompt buffer by hooking it to prompt-buffer-mode-after-hook, but it only turns on vi-insert-mode globally with the prompt still on vi-normal-mode.

This works: 
```
(define-configuration prompt-buffer
    ((default-modes (append '(vi-insert-mode) %slot-default%))))
```
I agree with you other suggestion!

**Comment by starryveshch** [comment]
> This works:
> 
> ```
> (define-configuration prompt-buffer
>     ((default-modes (append '(vi-insert-mode) %slot-default%))))
> ```
Wow, thanks! I never thought it would be this simpler, but I still wish it is something present by default.

**Comment by edgar-vincent** [comment]
I think so too. I mentioned this in this issue: https://github.com/atlas-engineer/nyxt/issues/1355#issuecomment-838361704

**Comment by alejandrogallo** [comment]
I think it would be a good idea in general to define
alongside the `normal` and `insert` mode
the `command` mode, which is the mode used in the prompt
in vim. Qutebrowser also has this, and I think the example config
for vi stuff should have this in the example. Right now I'm starting to configure
so I'm also encountering these kind of problems.

For instance in qutebrowser I have the following

```python
imap = lambda x, y: config.bind(x, y, mode="insert")
cmap = lambda x, y: config.bind(x, y, mode="command")
nmap = lambda x, y: config.bind(x, y, mode="normal")

[...]
cmap("<Ctrl+i>", "completion-item-focus next")
cmap("<Ctrl+f>", "edit-command")
cmap("<Ctrl+g>", "leave-mode")
cmap("<Ctrl+j>", "completion-item-focus next")
cmap("<Ctrl+k>", "completion-item-focus prev")
cmap("<Ctrl+d>", "repeat 10 completion-item-focus next")

[...]
imap("<Ctrl+c>", "fake-key <Esc>")
imap("<Ctrl+h>", "fake-key <backspace>")
imap("<Ctrl+j>", "fake-key <enter>")
imap("<Ctrl+u>", "fake-key <Shift+Home> ;; fake-key <Delete>")
imap("<Ctrl+w>", "fake-key <ctrl+backspace>")
imap("<Ctrl+g>", "fake-key <esc>")
imap("<Ctrl+i>", "fake-key <tab>")
imap("<Ctrl+m>", "fake-key <enter>")
```
and I'm really missing some of these behaviours in nyxt.
For instance, it is clear that in the prompt a vi user will want
`Tab`/`C-i` to autocomplete, and `C-w` to delete word,
`C-u` to delete the line to the beginning etc, but probably not in insert mode.

**Comment by alejandrogallo** [comment]
> > This works:
> > ```
> > (define-configuration prompt-buffer
> >     ((default-modes (append '(vi-insert-mode) %slot-default%))))
> > ```
> 
> Wow, thanks! I never thought it would be this simpler, but I still wish it is something present by default.

this should at least be in the documentation manual.

**Comment by dev-zero** [comment]
The `prompt-buffer` configuration is only for ensuring `vi-insert-mode` for the `prompt`, there is no solution/workaround yet for automatically switching to insert mode when entering an input field on a webpage, correct?

**Comment by starryveshch** [comment]
@dev-zero  Right now, the automatic switching works when clicking directly on the input fields or when focusing on the first input field using `M-i`. It only doesn't work when entering input fields through hinting.

**Comment by dev-zero** [comment]
@starryveshch I see, thanks for the clarification.

**Comment by Ambrevar** [comment]
Should be fixed by #1499.
